### PR TITLE
Stop using RTLD_GLOBAL for _C

### LINKED ISSUE #3059: Avoid loading with RTLD_GLOBAL
It seems PyTorch loads the .so using RTLD_GLOBAL which exports many symbols into shared namespace. This can causes crashes when another library is using different version of the symbol with the same name.

In particular, doing `import torch; import tensorflow` will crash because pytorch .so is exposing `pthread` symbols
https://github.com/tensorflow/tensorflow/issues/13615#issuecomment-335641777

TensorFlow recently switched away from RTLD_GLOBAL in https://github.com/tensorflow/tensorflow/commit/5c7f9e31

### PR DESCRIPTION
**Pull request by apaszke**
This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

One complication is that MKL doesn't like being in a local scope, unless
you use the libmkl_rt.so variant, so I adapted our CMake configuration
to either accept that or only allow static linking.

All relocations are also resolved eagerly now, which should let us avoid
some bugs. In the process of writing this patch I found e.g. a missing
TORCH_API declaration in the NCCL bindings.

We also no longer use ctypes to interface with CUDA libraries, but use
pybind to expose the relevant functionality. This has the upside of
no longer forcing us to hardcode the enum values in Python, and also
should let us remove the ungodly hacks we use to load the libraries on
Windows.

Finally, turns out that most of our cuDNN interface in Python is dead
now so I have removed it.

Fixes #3059.

### CODE REVIEW

File: cmake/Modules/FindMKL.cmake [hunk 1]
```diff
@@ -271,7 +278,19 @@ IF (NOT "${MKL_THREADING}" STREQUAL "SEQ")
...
+FOREACH(mklrtl ${mklrtls} "")
+  FOREACH(mkliface ${mklifaces})
+    FOREACH(mkl64 ${mkl64s} "")
+      IF (NOT MKL_LIBRARIES)
+        CHECK_ALL_LIBRARIES(MKL_LIBRARIES MKL_OPENMP_TYPE MKL_OPENMP_LIBRARY cblas_sgemm
+          "mkl_${mkliface}${mkl64};${mklthread};mkl_core;${mklrtl};pthread;${mkl_m};${mkl_dl}" "")
+      ENDIF (NOT MKL_LIBRARIES)
+    ENDFOREACH(mkl64)
+  ENDFOREACH(mkliface)
+ENDFOREACH(mklrtl)
```
Comment (apaszke) [inline cmake/Modules/FindMKL.cmake]: I've moved the pthread implementations to take priority over the sequential ones. I'm not sure why has it been that way previously.

File: cmake/Modules/FindMKL.cmake [hunk 2]
```diff
@@ -333,23 +340,10 @@ IF (MKL_LIBRARIES)
...
-  IF (CMAKE_COMPILER_IS_GNUCC)
-    FIND_LIBRARY(MKL_KERNEL_libirc "irc")
-  ELSEIF (CMAKE_C_COMPILER_ID AND NOT CMAKE_C_COMPILER_ID STREQUAL "Intel")
-    FIND_LIBRARY(MKL_KERNEL_libirc "irc")
-  ENDIF (CMAKE_COMPILER_IS_GNUCC)
-  MARK_AS_ADVANCED(MKL_KERNEL_libirc)
-  IF (MKL_KERNEL_libirc)
-    SET(MKL_LIBRARIES ${MKL_LIBRARIES} ${MKL_KERNEL_libirc})
-  ENDIF (MKL_KERNEL_libirc)
-ENDIF (MKL_LIBRARIES)
```
Comment (apaszke) [inline cmake/Modules/FindMKL.cmake]: I can't find any reasonable references to libIRC online, so I decided to remove this.

File: setup.py [hunk 3]
```diff
@@ -607,29 +607,11 @@ def configure_extension_build():
...
-    # and libcaffe2.so was brought in as a dependency (though I suspect it
-    # may have been possible to break by importing caffe2 first in the
-    # same process).
-    #
-    # Now, libtorch_python.so and libcaffe2.so define some of the same
-    # symbols. We directly load the _C.so stub, which brings both of these
-    # in as dependencies. We have to make sure that symbols continue to be
-    # looked up in libtorch_python.so first, by making sure it comes
-    # before libcaffe2.so in the linker command.
-    main_link_args.extend(CAFFE2_LIBS)
```
Comment (apaszke) [inline setup.py]: This seems like an irrelevant thing now, so should be safe to remove.

File: torch/__init__.py [hunk 4]
```diff
@@ -33,61 +33,28 @@
...
-try:
-    import numpy as _np  # noqa: F401
-except ImportError:
-    pass
-
 if platform.system() == 'Windows':
-    # first get nvToolsExt PATH
-    def get_nvToolsExt_path():
-        NVTOOLEXT_HOME = _dl_flags.getenv('NVTOOLSEXT_PATH', 'C:\\Program Files\\NVIDIA Corporation\\NvToolsExt')
+    NVTOOLSEXT_PATH = os.getenv('NVTOOLSEXT_PATH', 'C:\\Program Files\\NVIDIA Corporation\\NvToolsExt')
```
Comment (apaszke) [inline torch/__init__.py]: @peterjc123 Do we really need this code now? We're no longer using `ctypes` to dlopen any libraries -- they should all get listed as dependencies of libtorch.

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -431,17 +431,17 @@ def CppExtension(name, sources, *args, **kwargs):
...
     if IS_WINDOWS:
-        library_dirs = kwargs.get('library_dirs', [])
-        library_dirs += library_paths()
-        kwargs['library_dirs'] = library_dirs
-
-        libraries = kwargs.get('libraries', [])
-        libraries.append('c10')
-        libraries.append('torch')
-        libraries.append('torch_python')
         libraries.append('_C')
```
Comment (apaszke) [inline torch/utils/cpp_extension.py]: I think this one should be safe to delete too, as `_C` is only a stub for `libtorch_python` now, but I want to get the Windows CI to pass before I do this cleanup too.

**Comment by apaszke** [comment]
Important question: do we ever use other threading libraries of MKL other than Intel (e.g. TBB)? If so, we might need some extra code to make `libmkl_rt` pick the right one at run time.

**Comment by soumith** [comment]
Fixes https://github.com/pytorch/pytorch/issues/19668
Fixes https://github.com/pytorch/pytorch/issues/3059

File: torch/__init__.py (hunk 4 above)
Comment (peterjc123) [reply torch/__init__.py]: <del> But the dynamic library `torch.dll` relies on `nvtoolext64_1.dll`, right? In that case, we still need it because the environmental variable is not reliable from the user's side. </del> We can package that dll into the package so that we don't really need this code.

File: torch/__init__.py [hunk 6]
```diff
@@ -33,61 +33,28 @@
...
 
-        if _dl_flags.path.exists(NVTOOLEXT_HOME):
-            return _dl_flags.path.join(NVTOOLEXT_HOME, 'bin', 'x64')
-        else:
-            return ''
+    if os.path.exists(NVTOOLEXT_HOME):
```
Comment (peterjc123) [inline torch/__init__.py]: Needs to be changed to `NVTOOLEXT_PATH` according to L37.

File: torch/utils/cpp_extension.py (hunk 5 above)
Comment (peterjc123) [reply torch/utils/cpp_extension.py]: Yes, I agree.

File: torch/cuda/__init__.py [hunk 7]
```diff
@@ -22,44 +22,17 @@
 from multiprocessing.util import register_after_fork as _register_after_fork
 from ._utils import _get_device_index
 
+try:
+    from torch._C import _cudart
+except ImportError:
+    pass
```
Comment (albanD) [inline torch/cuda/__init__.py]: Why is that ok to ignore? Won't the error message be quite unfriendly if something in this file tries to use `_cudart` like to check a cuda error? (same for _cudnn and _nvtx)

File: torch/backends/cudnn/rnn.py [hunk 8]
```diff
@@ -1,16 +1,16 @@
 import torch.cuda
-import torch.backends.cudnn as cudnn
+from torch._C import _cudnn
```
Comment (albanD) [inline torch/backends/cudnn/rnn.py]: Is this missing a try/catch ?

File: torch/csrc/Module.cpp [hunk 9]
```diff
@@ -709,6 +717,11 @@ PyObject* initModule() {
   torch::python::init_bindings(module);
 #ifdef USE_CUDA
   torch::cuda::initModule(module);
+  torch::cuda::shared::initCudartBindings(module);
```
Comment (albanD) [inline torch/csrc/Module.cpp]: Could these be part of the cuda initModule? Or you think it's clearer to add them here explicitly?

**Comment by gchanan** [comment]
> Important question: do we ever use other threading libraries of MKL other than Intel (e.g. TBB)? If so, we might need some extra code to make libmkl_rt pick the right one at run time.

CC @ilia-cher.

File: torch/cuda/__init__.py (hunk 7 above)
Comment (apaszke) [reply torch/cuda/__init__.py]: It's ok to ignore, because it's not the case that it will always be there (e.g. in CPU builds). But I agree that we should be careful about error messages.

File: torch/backends/cudnn/rnn.py (hunk 8 above)
Comment (apaszke) [reply torch/backends/cudnn/rnn.py]: Good catch 😉

File: torch/csrc/Module.cpp (hunk 9 above)
Comment (apaszke) [reply torch/csrc/Module.cpp]: I guess I could do that too, no strong preference.

**Comment by ilia-cher** [comment]
> > Important question: do we ever use other threading libraries of MKL other than Intel (e.g. TBB)? If so, we might need some extra code to make libmkl_rt pick the right one at run time.
> 
> CC @ilia-cher.

" threading libraries of MKL other than Intel (e.g. TBB)?" - not sure what you mean by this, Intel is providing MKL, TBB, as well as a version of MKL that uses TBB

And, yes, we do use MKL-TBB - please refer to https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html

Make sure that TBB (and native too) build is working, I added a TBB CI build on master in https://github.com/pytorch/pytorch/pull/27925

**Comment by apaszke** [comment]
Eh turns out that many builds of OpenMPI [assume they're loaded in the global namespace, which trips up most of our CI runs](https://www.open-mpi.org/faq/?category=troubleshooting) 🤦‍♂

File: torch/__init__.py (hunk 4 above)
Comment (apaszke) [reply torch/__init__.py]: Can you elaborate? Why would they have issues?

Comment (peterjc123) [reply torch/__init__.py]: Because:
1. The directory of the nvtx library is not added into PATH on Windows. Windows relies on PATH to search for dynamic libraries.
2. All the dynamic libraries are eagerly loaded (not lazily) unless `/DELAYLOAD` is specified on the linker line. https://docs.microsoft.com/en-us/cpp/build/reference/delayload-delay-load-import?view=vs-2019
3. The following dependency chain exists: `_C -> torch_python -> nvtx`, so I don't think it's only a dependent of libtorch. 
https://github.com/pytorch/pytorch/pull/28536/files#diff-2eeaed663bd0d25b7e608891384b7298L613 https://github.com/pytorch/pytorch/pull/28536/files#diff-94eab65863273226cd6b960d685a427eR179

When users compile PyTorch by themselves, they don't package the nvtx library along with the torch libraries. (which is done through the release scripts in pytorch/builder). Then they will experience the import errors because `nvtoolsext64_1.dll` could not be found at runtime.

Comment (apaszke) [reply torch/__init__.py]: So basically what you're saying is that there's nothing like RPATH for Windows, and all the DLLs that you depend on should be in your PATH when you load a shared library?

Comment (peterjc123) [reply torch/__init__.py]: Yeah, that's what I mean. For more details, you may read [this](http://ibob.github.io/blog/2018/12/16/windows-rpath/) article.

**Comment by apaszke** [comment]
@ilia-cher it seems like the TBB CI is not triggered in this PR, any clue for why would that be?

**Comment by ilia-cher** [comment]
@suo @ezyang preferred TBB CI builds to be triggered on master only 

please follow https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html on how to build it locally, e.g. TBB build can be built with
USE_TBB=1 BLAS=MKL MKL_THREADING=TBB ATEN_THREADING=TBB USE_MKLDNN=1 MKLDNN_THREADING=TBB 

Also please add BUILD_BINARY=1 to build `parallel_info` utility - output of this utility (that says that TBB is used) + passing tests should be enough

**Comment by ilia-cher** [comment]
We also need to verify that `MKL_THREADING=SEQ` results in sequantial mkl library loaded

**Comment by ezyang** [comment]
Some notes on my debugging progress for the remaining CI failures on this patch:
* I created a stack diff, top of which is https://github.com/pytorch/pytorch/pull/31162 , to factor the various pieces of this PR into smaller chunks to make it easier to localize failures. @apaszke if you could take a quick look and agree that my separation is logical, that would be helpful. The commits are not attributed correctly yet but hopefully I'll remember to fix this before landing (won't happen until new year, anyway)
* The changes to the MKL cmake appear to be responsible for precision problems on some autograd tests, see https://github.com/pytorch/pytorch/pull/31159  . I had originally thought that this was related to preferring the parallel mkl implementation over the sequential one, but subsequent testing reveals that the precision error happens without that reordering. The failure looks like:

```
Dec 12 05:03:50 ======================================================================
Dec 12 05:03:50 FAIL [0.544s]: test_grad_conv3d_weight (__main__.TestNN)
Dec 12 05:03:50 ----------------------------------------------------------------------
Dec 12 05:03:50 Traceback (most recent call last):
Dec 12 05:03:50   File "test_nn.py", line 7708, in test_grad_conv3d_weight
Dec 12 05:03:50     self.run_grad_conv_test(F.conv3d, F.grad.conv3d_weight, 3, 'weight')
Dec 12 05:03:50   File "test_nn.py", line 7690, in run_grad_conv_test
Dec 12 05:03:50     dilation=dilation))
Dec 12 05:03:50   File "/var/lib/jenkins/workspace/test/common_utils.py", line 847, in assertAlmostEqual
Dec 12 05:03:50     self.assertEqual(x, y, prec, msg, allow_inf)
```

* The MKL cmake changes are also causing this problem on OS X:

```
Dec 11 20:51:21 ++ python -c 'import torch; print(int(torch.backends.openmp.is_available()))'
Dec 11 20:51:22 OMP: Error #15: Initializing libomp.dylib, but found libiomp5.dylib already initialized.
Dec 11 20:51:22 OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/
Dec 11 20:51:22 + [[ ! '' == \1 ]]
Dec 11 20:51:22 + echo 'Build should have OpenMP enabled, but torch.backends.openmp.is_available() is False'
Dec 11 20:51:22 Build should have OpenMP enabled, but torch.backends.openmp.is_available() is False
```

* There's an AMD build problem because we aren't HIPifying the header correctly

* The ASAN failure says that we are somehow ending up with multiple vtables for `AutogradMeta` in our address space. This might be "intentional" on the part of this PR: if we load `_C.so` with `RTLD_LOCAL`, that means the `AutogradMeta` symbol from `libtorch` won't be used for relocations. If we load `libtorch.so` again later when loading a C++ extension, we'll end up with our second copy. I'm currently checking `LD_DEBUG` log output to see if this is what actually happened.

**Comment by ezyang** [comment]
> One complication is that MKL doesn't like being in a local scope, unless
> you use the libmkl_rt.so variant, so I adapted our CMake configuration
> to either accept that or only allow static linking.

@apaszke But then, later in this patch, you add an extra library to suck the MKL symbols into global scope anyways; this part of the diff:

```
+# See Note [Global dependencies]
+def _load_global_deps():
+    if platform.system() == 'Windows':
+        return
+    lib_name = 'libtorch_global_deps' + ('.dylib' if platform.system() == 'Darwin' else '.so')
+    here = os.path.abspath(__file__)
+    lib_path = os.path.join(os.path.dirname(here), 'lib', lib_name)
+
+    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
+
+
+# See Note [Global dependencies]
+_load_global_deps()
```

If you're doing this unconditionally, do you really need to also make the MKL cmake changes? (Because those changes seem to be introducing bugs!!)

**Comment by ezyang** [comment]
Well, fuck.

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/shared_ptr_base.h:752:11: runtime error: member call on
 address 0x6030003177d0 which does not point to an object of type 'std::_Sp_counted_base<__gnu_cxx::_Lock_policy::_S_atomic>'
0x6030003177d0: note: object is of type 'std::_Sp_counted_ptr<torch::nn::LinearImpl*, (__gnu_cxx::_Lock_policy)2>'
 ee 01 80 57  b0 82 3d fa 86 7f 00 00  00 00 00 00 02 00 00 00  80 5a 4b 00 50 61 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by ezyang** [comment]
Updates:
* I'm "fixing" the UBSAN problem by just preloading libtorch.so, so that the weak symbols get resolved consistently. RTTI isn't going to work with extensions but I'm going to chalk this up as "by design"
* I hypothesize that the `mkl_rt` changes aren't needed, and all we need to do is just preload MKL with `RTLD_GLOBAL` and then libtorch will load just fine. I'm hoping this fixes the 2.7.9-only numeric precision problems. My current experiment to do this is not working (https://github.com/pytorch/pytorch/pull/31162) failing with:

```
Dec 13 03:50:14 /opt/python/2.7.9/bin/python: symbol lookup error: /opt/python/2.7.9/lib/libmkl_vml_avx2.so: undefined symbol: mkl_serv_getenv
```

* I feel that the ctypes changes are unnecessary. So I've also removed them from the PR (additionally, now I don't have to fix peterjc123's comments about how we woudl have broken windows)

File: torch/__init__.py (hunk 4 above)
Comment (peterjc123) [reply torch/__init__.py]: Actually, things are getting worse in Python 3.8. It doesn't read from `PATH` anymore, but os.add_dll_directory should be called manually instead. This means that we will need to add more code here, for finding cuda and cudnn DLLs. In order to make this file clean, maybe we should move this into a different script in the root directory.

**Comment by ezyang** [comment]
More updates:
* I'm also going to preload the CUDA libraries so that I can get rid of all of the work Adam did to stop using ctypes here.

**Comment by ezyang** [comment]
Final update: #31162 seems to be passing all tests and ready to go
